Contest_API.md

```diff
@@ -116,27 +116,36 @@ their expected behavior, if implemented.
     It can be used to request a whole collection or a specific element.
   - `POST`
     Create a new element. This can only be called on a collection
-    endpoint. No `id` attribute should be specified as it is up to the
-    server to assign one, which is returned in the location header.
+    endpoint. When using most [roles](#Roles) the `id` attribute may not
+    be specified as it is up to the server to assign one. When using an
+    `admin` role you may specify a new, unused id.
```

Comment:
1. If you specify an `id` in your `POST`, you could have used a `PUT` instead, and as a general rule one should prefer `PUT` over `POST`, since `PUT` is idempotent, and `POST` is not. (This general rule is definitely being argued on the Internet though...)
One could make the case that the difference between `POST`ing with an `id` and `PUT`ing to the same `id` is that the former is a suggestion and the latter a requirement. `PUT`ing to an `id` should only ever either create or update that `id` or be rejected. `POST`ing *could* (and maybe should, I'm not sure?) create a different `id`.

2. Why are we limiting to "a new, unused id". In fact, using `POST` to update an existing `id` would make as much sense.

3. It's not true that you always may specify an `id` (unused or otherwise) when you are `admin`, it depends on the endpoint.

In summary:

I think that I think that whenever you are supplying an `id` with a `POST` it would be better to instead use the equivalent `PUT`. But that doesn't mean that we couldn't *also* allow the `POST`.

I think we should say less here. Maybe something like:
> Typically the `id` attribute must not be specified as it is up to the server to assign one.

Reply:
1. So you are also suggesting to change POST to `submissions` to be a PUT to `submissions/<id>` if you supply an ID? This does seem to be more in line with the RFC, but I am guessing that if we go this way that DOMjudge will just accept post POST and PUT to both the collection as well as a specific ID (and only verify if the ID matches anything you supply in the body
2. I am not a fan at all of updating the ID of an existing entity. How would that go in the event feed? This also means you need to propagate the ID to any entities that uses it (in the event feed). Please let's not allow this. Also, should this then not be a `PUT`?
3. I agree with your third point

Reply:
1/2. When we originally wrote this section I think the intent was clearly that POST was for creating an element and PUT was for replacing an existing element. I'm ok changing that and your points about the HTTP spec in 1 and 2 are correct, but you're also pointing how many ways POST and PUT are the same and I don't really want to make these methods basically equivalent or interchangeable either.

Reading the HTTP spec, the difference I get is POST is 'could you please accept this resource somewhere' and PUT is 'I want this resource here'. What if we left POST for cases where you don't have or know the id and it must not be included (e.g. teams submitting), and PUT for cases where you know the id and just want to set or replace the object (e.g. admins or some of the proxy cases)? To me this would provide some good differentiation in purpose and clarity to the behaviour.

If we went that way it would make total sense to not allow id in POST, require it in PUT (since you are replacing a 'whole' object), and have it optional in PATCH (since you only need to include any attributes you want to change).

As Nicky said I think PUT and PATCH should fail if the id in the url and attribute don't match, as that is changing the URL of the resource. It's simpler to just require a DELETE and then a PUT.

3. Agreed, I worded this badly.

Reply:
>     1. DOMjudge will just accept post POST and PUT to both the collection as well as a specific ID

Accepting a `PUT` to the collection is absolutely wrong (unless you're `PUT`ing the actual entire collection). Accepting a `POST` to the entity is arguably wrong. So, please don't do that.

>     2. I am not a fan at all of updating the ID of an existing entity.

Sorry, I was being very unclear. You should *not* be able to update the `id` of en existing object, what I meant by "update an existing `id`" was to update the object identified by some `id`.

> If we went that way it would make total sense to not allow id in POST, require it in PUT (since you are replacing a 'whole' object), and have it optional in PATCH (since you only need to include any attributes you want to change).

That sounds good to me.

> As Nicky said I think PUT and PATCH should fail if the id in the url and attribute don't match, as that is changing the URL of the resource. It's simpler to just require a DELETE and then a PUT.

Definitely agreed.

Reply:
> > ```
> > 1. DOMjudge will just accept post POST and PUT to both the collection as well as a specific ID
> > ```
> 
> 
> Accepting a `PUT` to the collection is absolutely wrong (unless you're `PUT`ing the actual entire collection). Accepting a `POST` to the entity is arguably wrong. So, please don't do that
PUTting to the collection is also wrong. So if we both promise to never to it we can’t verify we do 😅. We might add a check
> 
> 
> > ```
> > 2. I am not a fan at all of updating the ID of an existing entity.
> > ```
> 
> 
> Sorry, I was being very unclear. You should _not_ be able to update the `id` of en existing object, what I meant by "update an existing `id`" was to update the object identified by some `id`.
👍
> 
> 
> > If we went that way it would make total sense to not allow id in POST, require it in PUT (since you are replacing a 'whole' object), and have it optional in PATCH (since you only need to include any attributes you want to change).
> 
> 
> That sounds good to me.
> 
> 
> > As Nicky said I think PUT and PATCH should fail if the id in the url and attribute don't match, as that is changing the URL of the resource. It's simpler to just require a DELETE and then a PUT.
> 
> 
> Definitely agreed.

Reply:
> PUTting to the collection is also wrong. So if we both promise to never to it we can’t verify we do 😅. We might add a check

Ah, I see what you're saying. Yes, that makes sense. :)